src/App.jsx

```diff
@@ -1,7 +1,101 @@
+import { useState } from "react";
 import "./App.css";
+import BottomModalLayout from "./components/BottomModalLayout/BottomModalLayout";
+import Header from "./components/common/Header/Header";
+import RestaurantListLayout from "./components/RestaurantListLayout/RestaurantListLayout";
+import RestaurantDetailModal from "./components/BottomModalLayout/RestaurantDetailModal/RestaurantDetailModal";
+import AddRestaurantModal from "./components/BottomModalLayout/AddRestaurantModal/AddRestaurantModal";
+import { useEffect } from "react";
 
 function App() {
-  return <h1>Self-Paced React</h1>;
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const [modalName, setModalName] = useState("");
+  const [selectedRestaurant, setSelectedRestaurant] = useState(null);
+  const [restaurants, setRestaurants] = useState([]);
+
+  useEffect(() => {
+    const fetchRestaurants = async () => {
+      const response = await fetch("http://localhost:3000/restaurants");
+      const data = await response.json();
+      setRestaurants(data);
+    };
+
+    fetchRestaurants();
+  }, []);
+
+  const openRestaurantDetailModal = ({ restaurant }) => {
+    setModalName("restaurantDetail");
+    setSelectedRestaurant(restaurant);
+    setIsModalOpen(true);
+  };
+
+  const onAddRestaurant = async (event) => {
+    event.preventDefault();
+    const form = new FormData(event.target);
+    const newRestaurant = {
+      id: Date.now().toString(),
+      name: form.get("name"),
+      description: form.get("description"),
+      category: form.get("category"),
+    };
+
+    const response = await fetch("http://localhost:3000/restaurants", {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(newRestaurant),
+    });
+
+    if (!response.ok) {
+      alert("레스토랑 추가에 실패했습니다.");
+      return;
+    }
+
+    setRestaurants([...restaurants, newRestaurant]);
```

Comment:
```suggestion
    setRestaurants((prevRestaurants) => [...prevRestaurants, newRestaurant]);
```

https://github.com/woowacourse/self-paced-react/pull/21#discussion_r2033583822
세오꺼 코멘트에 남긴 것처럼 `setState`할 때 함수형 업데이트를 사용하면 이전 상태 값을 안전하게 가져올 수 있을 것 같아요!🙃

Reply:
역시 써밋...리액트 좀 치시네요...멋지당...✨

글 읽다보니, 개념이 클로저함수의 상태관리 원리와 유사해보이는데 그럼 클로저 함수처럼 메모리 누수 위험이 있을까? 싶어서 글을 더 찾아봤어요! 결론은 useState 훅은 ```메모리 누수의 위험이 없다!``` 네요.

>“각 컴포넌트와 관련된 메모리 셀의 내부 목록이 있고, useState는 그 셀에 값을 저장하고 포인터를 다음 셀로 이동한다.”

라는 말을 보면 useState는 렌더링될 때마다 React가 관리하는 ```메모리 셀 목록```을 기준으로 동작하기 때문에, 
이전 상태를 계속 참조하는걸로 GC에 들어가지 않게 하는 클로저랑은 동작이 엄밀히 다르고,

React가 컴포넌트 생명주기와 함께 상태를 정확하게 관리하기 때문에 메모리 해제가 되지 않는 누수가 발생하진 않는다는 것 같네요!

https://ko.legacy.reactjs.org/docs/hooks-faq.html#how-do-lifecycle-methods-correspond-to-hooks
https://ko.legacy.reactjs.org/docs/hooks-faq.html#how-does-react-associate-hook-calls-with-components


하지만 useEffect 훅을 사용할 때는 정리할 타이밍을 고려해서 적절히 clean-up 관리도 해주어야겠군요!
https://ko.legacy.reactjs.org/docs/hooks-effect.html

써밋덕분에 좋은 학습이 되었어요!!👍
공식 문서 타고타고 들여다보는 게 처음이라 살짝 헤맸는데, 레벨 2때는 공식문서 잘 쓰는 법(?)도 익혀야겠어요 🥲

Reply:
허거덩.. 저는 메모리 누수에 대해 생각해본 적도 없는데 대단하시네요...😮

![스크린샷 2025-04-10 오후 4 16 11](https://github.com/user-attachments/assets/e682b801-7161-41a1-8ca0-3cf204f78ef6)

메모리 누수의 관점에서는 잘 모르겠지만 저 내용은 리액트 훅 객체를 연결 리스트 형태로 관리하는 것과 관련이 있어보이네요!🤔

리액트 훅 객체에 `memoizedState`, `baseState` 이런 속성들이 있는데 아마 함수형 업데이트에서는 `baseState` 이 값을 기준으로 이전 상태 값을 참조하는 걸로 알고 있어여(아마도?) 자세한건 딥 다이브 스터디때 같이 알아보시져~~🙃🙃🙃

Reply:
허거덩거허 벌써 머리아프아요우오우요후를레이
딥 다이브 스터디에서 써교수님이 떠먹여주세요🙇🏻‍♀️ㅋㅎㅎㅎㅎ

Reply:
에이 메모리 누수까지 생각하시는 분이..ㅋㅋㅋㅋㅋ

오히려 제가 기대하고 있겠습니다🙇‍♂️